source/_posts/2017/2017-01-31-how-monolithic-repository-in-open-source-saved-my-lazyness.md

```diff
@@ -0,0 +1,159 @@
+
+It is used by:
+
+- [Nette](https://github.com/nette/)
+- [Doctrine](https://github.com/doctrine)
+- [PHPLeague](https://github.com/thephpleague)
+
+and almost everybody else.
+
+
+So how does this monorepo work?
+
+
+## From One Hearth to Many Arteries
+
+Imagine flow of a oxygenated blood from heart to arteries. All blood that was in heart, will drift to all arteries.
+
+<img src="/assets/images/posts/2017/monorepo/blood-vein.png" alt="Blod vein" class="thumbnail">
+
+From **1 heart** to **many arteries** in one direction. Same relation is between **1 monolithic repository** and **many repositories**.
+
+
+This is related to commits and tagging:
+
+- **Every commit** that does changes is Symfony\Console code in monorepo, is also in Symfony\Console manyrepo.
+- **Every tag** in monorepo, is also in all manyrepos.
+
+
+## What Are top 4 Advantages of Monorepo?
+
+### 1. It scales
+
+You have one repo to maintain.
+
+- New package? Not a problem.
+- Changing name of method that is used by 10 other packages? Easier than ever.
+- Starting a framework? Ideal.
+
+That's the way Nette, Symfony and Laravel started and grow so much, even if only one person was behind majority of commits.
+
+
+### 2. Upgrade Fast, Upgrade Safe
+
+It's easy to make changes in that affect all packages - bump to PHP 7.1, Symfony 3.2 or Nette 3.0 is matter of minutes.
+And I know it works on all packages. Before monorepo, I had to upgrade every package manually, which resulted in dissonance:
+one package used Symfony\Console 3.2, but other only 2.8 and it got messy for no reason.
+
+
+### 3. Test Both on Monorepo and Manyrepo Level With Ease
+
+Another thing I love is testing both monorepo (all packages together) and manyrepo level. Once I spend 5 hours over bug in Symfony\Process. It was difficult to find, because all tests were passing. Even for testing just the Symfony\Process directory made it pass. But when I copied only the Symfony\Process code to the standalone directory, it failed. Few hours later I found out, it's somehow related to having Symfony\Stopwatch presenter. Yea, WTF.
+
+**That could be caught testing on monorepo level.**
+
+On the other hand, monorepo testing is also important. When Nette [was split from monorepo to manyrepo only](https://phpfashion.com/prave-jsem-smazal-nette-framework), all tests were passing packages were standalone. But in combination some of them didn't.
+
+This is not issue of the code itself, but of the testing architecture.
+
+
+### 4. The Burnout is Much More Harder
+
+When maintaining 15 own packages, [ApiGen](https://github.com/Apigen) and co-maintaining few more repositories, I spent a lot of time by package management and not coding. It wasn't fun and I contributed less and less.
+
+Many packages like [Doctrine](http://github.com/doctrine) or [Kdyby](https://github.com/Kdyby) are slowing down in evolution because of this.
+
+Again, it's rather matter of project  architecture, not the code or maintainers.
```

Comment:
Again, it's a matter of project architecture rather than the code or maintainers.